lib/spack/spack/bootstrap.py

```diff
@@ -28,6 +31,208 @@
+                        "modules": [],
+                        "operating_system": str(index_spec.os),
+                        "paths": {
+                            "cc": "/dev/null",
+                            "cxx": "/dev/null",
+                            "f77": "/dev/null",
+                            "fc": "/dev/null"
+                        },
+                        "spec": str(index_spec.compiler),
+                        "target": str(index_spec.target)
+                    }
+                    with spack.architecture.use_platform(bincache_platform):
+                        with spack.config.override(
+                                'compilers', [{'compiler': compiler_entry}]
+                        ):
+                            spec_str = '/' + pkg_hash
+                            install_args = [
+                                'install',
+                                '--sha256', pkg_sha256,
+                                '-a', '-u', '-o', '-f', spec_str
+                            ]
+                            buildcache(*install_args, fail_on_error=False)
+                # TODO: undo installations that didn't complete?
+
+                if _StoreBootstrapper.try_import(module, abstract_spec_str):
+                    return True
+        return False
+
+
+@_bootstrapper(type='install')
+class _SourceBootstrapper(object):
+    def __init__(self, conf):
+        self.conf = conf
+
+    @staticmethod
+    def try_import(module, abstract_spec_str):
+        # Try to build and install from sources
+        with spack_python_interpreter():
+            # Add hint to use frontend operating system on Cray
+            if str(spack.architecture.platform()) == 'cray':
+                abstract_spec_str += ' os=fe'
+
+            concrete_spec = spack.spec.Spec(
+                abstract_spec_str + ' ^' + spec_for_current_python()
+            )
+            # TODO: substitute this call when the old concretizer is deprecated
+            concrete_spec._old_concretize()
+
+        msg = "[BOOTSTRAP MODULE {0}] Try installing '{1}' from sources"
+        tty.debug(msg.format(module, abstract_spec_str))
+
+        # Install the spec that should make the module importable
+        concrete_spec.package.do_install()
+
+        return _StoreBootstrapper.try_import(
+            module, abstract_spec=abstract_spec_str
+        )
+
+
+def _make_bootstrapper(conf):
+    """Return a bootstrap object built according to the
+    configuration argument
+    """
+    btype = conf['type']
+    return _bootstrap_methods[btype](conf)
+
+
+def _source_is_trusted(conf):
+    trusted, name = spack.config.get('bootstrap:trusted'), conf['name']
+    # We always trust something that is already installed. Since
+    # "bootstrap-store" does not appear in the config, but is added
+    # in the code, we add the corresponding entry here
+    trusted['bootstrap-store'] = True
+    if name not in trusted:
+        return False
```

Comment:
This should prompt the user when the value is unset, and fill in the appropriate value. That way spack works "out of the box" still, without having to trust anything potentially scary. The user will have to decide which bootstrappers to trust on first invocation.

Reply:
For trusting I have mixed feelings. For interactive use prompting the user when something is not set it's possibly the best thing, but it may cause scripted use of Spack to hang indefinitely waiting for an answer. 

I don't think we can just distinguish with `tty` vs. not `tty` since CI would generally hang. With @tgamblin we were thinking of adding a warning message if something is in unknown state to tell the user that they can either trust or untrust that method. Would this solution be fine with you or do you see issues with that which were not considered?

Reply:
I would prefer an option to turn off prompting (or maybe a prompt with a timeout?) over a warning.

I think if the default experience for a user is `clone, install, it takes forever to bootstrap clingo` then we will lose a lot of people, and bootstrapping from source is less likely to work than bootstrapping from binary, because the binary only requires the interpreter but bootstrapping from source requires libpython and the dev headers.